apps/prairielearn/src/pages/instructorFileTransfer/instructorFileTransfer.ts

```diff
@@ -46,41 +71,60 @@ router.get(
-    try {
-      await editor.executeWithServerJob(serverJob);
-    } catch {
-      res.redirect(res.locals.urlPrefix + '/edit_error/' + serverJob.jobSequenceId);
-      return;
-    }
 
-    debug('Soft-delete file transfer');
-    await sqldb.queryOneRowAsync(sql.soft_delete_file_transfer, {
-      id: req.params.file_transfer_id,
-      user_id: res.locals.user.user_id,
-    });
-    debug(`Get question_id from uuid=${editor.uuid} with course_id=${res.locals.course.id}`);
-    const question = await selectQuestionByUuid({
-      // TODO: we'll have to change something here once we allow instructors to
-      // copy questions that have been shared with their course.
-      course_id: res.locals.course.id,
-      uuid: editor.uuid,
-    });
-    flash(
-      'success',
-      'Question copied successfully. You are now viewing your copy of the question.',
-    );
-    res.redirect(`${res.locals.urlPrefix}/question/${question.id}/settings`);
+    if (file_transfer.transfer_type === 'CopyQuestion') {
+      const qid = getContentFolder(file_transfer.from_filename, 'questions');
+      const editor = new QuestionTransferEditor({
+        locals: res.locals as any,
+        from_qid: qid,
+        from_course_short_name: from_course.short_name,
+        from_path: path.join(config.filesRoot, file_transfer.storage_filename),
+      });
+
+      await doTransfer(res, editor, file_transfer.id);
+
+      const question = await selectQuestionByUuid({
+        course_id: res.locals.course.id,
+        uuid: editor.uuid,
+      });
+
+      flash(
+        'success',
+        'Question copied successfully. You are now viewing your copy of the question.',
+      );
+      res.redirect(`${res.locals.urlPrefix}/question/${question.id}/settings`);
+    } else if (file_transfer.transfer_type === 'CopyCourseInstance') {
+      const course = await selectCourseById(file_transfer.from_course_id);
+      const shortName = getContentFolder(file_transfer.from_filename, 'courseInstances');
+
+      const fromCourseInstance = await selectCourseInstanceByShortName({
+        course_id: file_transfer.from_course_id,
+        short_name: shortName,
+      });
+
+      const editor = new CourseInstanceCopyEditor({
+        locals: res.locals as any,
+        from_course: course,
+        from_path: path.join(config.filesRoot, file_transfer.storage_filename),
+        course_instance: fromCourseInstance,
+      });
+
+      await doTransfer(res, editor, file_transfer.id);
+
+      const courseInstanceId = await selectCourseInstanceIdByUuid({
+        uuid: editor.uuid,
+        course_id: res.locals.course.id,
+      });
+
+      flash(
+        'success',
+        'Course Instance copied successfully. You are now viewing your copy of the Course Instance.',
+      );
+      // Redirect to the copied course instance
+      res.redirect(
+        `${res.locals.plainUrlPrefix}/course_instance/${courseInstanceId}/instructor/instance_admin/assessments`,
+      );
+    }
```

Comment:
Nit, consider adding this, I like having TypeScript perform exhaustiveness checking for us:

```suggestion
    } else {
      assertNever(file_transfer.transfer_type);
    }
```

`assertNever` is this:

https://github.com/PrairieLearn/PrairieLearn/blob/5e809a8b3fd9e4f3707c7e5afdab81f0d725815c/apps/prairielearn/src/lib/types.ts#L51-L53

Reply:
done